server/routing/pubsub/manager.go

```diff
@@ -0,0 +1,282 @@
+		ctx:         ctx,
+		host:        h,
+		pubsub:      ps,
+		topic:       topic,
+		sub:         sub,
+		localPeerID: h.ID().String(),
+		topicName:   TopicLabels,
+	}
+
+	// Start message handler goroutine
+	go manager.handleMessages()
+
+	logger.Info("GossipSub manager initialized",
+		"topic", TopicLabels,
+		"maxMessageSize", MaxMessageSize,
+		"peerID", manager.localPeerID)
+
+	return manager, nil
+}
+
+// PublishLabels announces labels for a record to the network.
+// This is called when a record is stored locally and should be
+// discoverable by remote peers.
+//
+// Flow:
+//  1. Convert labels.Label to wire format ([]string)
+//  2. Create and validate LabelAnnouncement
+//  3. Publish to GossipSub topic
+//  4. GossipSub mesh propagates to all subscribed peers
+//
+// Parameters:
+//   - ctx: Context for operation timeout/cancellation
+//   - cid: Content ID of the record
+//   - labelList: List of labels extracted from the record
+//
+// Returns:
+//   - error: If validation or publishing fails
+//
+// Note: This is non-blocking. GossipSub handles propagation asynchronously.
+func (m *Manager) PublishLabels(ctx context.Context, cid string, labelList []labels.Label) error {
+	// Convert labels.Label to strings for wire format
+	labelStrings := make([]string, len(labelList))
+	for i, label := range labelList {
+		labelStrings[i] = label.String()
+	}
+
+	// Create announcement with current timestamp
+	announcement := &LabelAnnouncement{
+		CID:       cid,
+		PeerID:    m.localPeerID,
+		Labels:    labelStrings,
+		Timestamp: time.Now(),
+	}
+
+	// Validate before publishing to catch issues early
+	if err := announcement.Validate(); err != nil {
+		return fmt.Errorf("invalid announcement: %w", err)
+	}
+
+	// Serialize to JSON
+	data, err := announcement.Marshal()
+	if err != nil {
+		return fmt.Errorf("failed to marshal announcement: %w", err)
+	}
+
+	// Publish to GossipSub topic
+	if err := m.topic.Publish(ctx, data); err != nil {
+		return fmt.Errorf("failed to publish announcement: %w", err)
+	}
+
+	logger.Info("Published label announcement",
+		"cid", cid,
+		"labels", len(labelList),
+		"topicPeers", len(m.topic.ListPeers()),
+		"size", len(data))
+
+	return nil
+}
+
+// SetOnLabelAnnouncement sets the callback for received label announcements.
+// This callback is invoked for each valid announcement received from remote peers.
+//
+// The callback should:
+//   - Convert wire format ([]string) to labels.Label
+//   - Build enhanced keys using BuildEnhancedLabelKey()
+//   - Store labels.LabelMetadata in datastore
+//
+// Example:
+//
+//	manager.SetOnLabelAnnouncement(func(ctx context.Context, ann *LabelAnnouncement) {
+//	    for _, labelStr := range ann.Labels {
+//	        label := labels.Label(labelStr)
+//	        key := BuildEnhancedLabelKey(label, ann.CID, ann.PeerID)
+//	        // ... store in datastore ...
+//	    }
+//	})
+func (m *Manager) SetOnLabelAnnouncement(fn func(context.Context, *LabelAnnouncement)) {
```

Comment:
likewise, in this method, we can register event listeners for record publications


```suggestion
func (m *Manager) SetRecordAnnouncementCallback(fn func(context.Context, *RecordPublishEvent)) {
```